File: backend/cleaning/cleaner.py
from datetime import datetime
from collections import OrderedDict

# Tracks recently seen events to deduplicate rapid-fire clicks
recent_events = OrderedDict()
DEDUP_WINDOW_MS = 50  # ignore identical clicks within 50ms

def clean_event(raw):
    """Validate and normalize incoming event data before storage"""
    errors = []
    
    # 1. Validate required fields
    if not isinstance(raw.get('user_id'), str) or not raw.get('user_id'):
        errors.append('Invalid user_id')
    if not isinstance(raw.get('session_id'), str) or not raw.get('session_id'):
        errors.append('Invalid session_id')
    if not isinstance(raw.get('x'), (int, float)) or not isinstance(raw.get('y'), (int, float)):
        errors.append('Invalid coordinates')
    
    # 2. Validate coordinate bounds
    if raw.get('x', -1) < 0 or raw.get('x', 10001) > 10000:
        errors.append('x coordinate out of range')
    if raw.get('y', -1) < 0 or raw.get('y', 10001) > 10000:
        errors.append('y coordinate out of range')
    
    if errors:
        return {'valid': False, 'errors': errors}
    
    # 3. Normalize timestamp — use server time if missing or invalid
    try:
        timestamp = datetime.fromisoformat(raw.get('timestamp', '')) if isinstance(raw.get('timestamp'), str) else datetime.now()
    except (ValueError, TypeError):
        timestamp = datetime.now()
    
    # 4. Deduplicate: same user + same pixel within 50ms window
    dedup_key = f"{raw['user_id']}:{int(raw['x'])}:{int(raw['y'])}"
    last_seen = recent_events.get(dedup_key)
    
    current_ms = timestamp.timestamp() * 1000
    if last_seen and current_ms - last_seen < DEDUP_WINDOW_MS:
        return {'valid': False, 'errors': ['Duplicate event (deduplicated)']}
    
    recent_events[dedup_key] = current_ms
    
    # Clean up old entries every 1000 events to prevent memory leak
    if len(recent_events) > 1000:
        cutoff = (datetime.now().timestamp() * 1000) - 5000
        keys_to_remove = [k for k, v in recent_events.items() if v < cutoff]
        for k in keys_to_remove:
            del recent_events[k]
    
    return {
        'valid': True,
        'cleaned': {
            'user_id': raw['user_id'],
            'session_id': raw['session_id'],
            'x': float(raw['x']),
            'y': float(raw['y']),
            'timestamp': timestamp,
            'page': raw.get('page', '/')
        }
    }

File: backend/cleaning/cleaner.js
/**
 * Data Cleaning Layer
 * Validates and normalizes incoming event data before storage.
 */

// Tracks recently seen events to deduplicate rapid-fire clicks
const recentEvents = new Map();
const DEDUP_WINDOW_MS = 50; // ignore identical clicks within 50ms

function cleanEvent(raw) {
    const errors = [];

    // 1. Validate required fields
    if (!raw.user_id || typeof raw.user_id !== 'string') errors.push('Invalid user_id');
    if (!raw.session_id || typeof raw.session_id !== 'string') errors.push('Invalid session_id');
    if (typeof raw.x !== 'number' || typeof raw.y !== 'number') errors.push('Invalid coordinates');

    // 2. Validate coordinate bounds (must be positive, reasonable screen size)
    if (raw.x < 0 || raw.x > 10000) errors.push('x coordinate out of range');
    if (raw.y < 0 || raw.y > 10000) errors.push('y coordinate out of range');

    if (errors.length > 0) {
        return { valid: false, errors };
    }

    // 3. Normalize timestamp — use server time if missing or invalid
    let timestamp = new Date(raw.timestamp);
    if (isNaN(timestamp.getTime())) {
        timestamp = new Date();
    }

    // 4. Deduplicate: same user + same pixel within 50ms window
    const dedupKey = `${raw.user_id}:${Math.round(raw.x)}:${Math.round(raw.y)}`;
    const lastSeen = recentEvents.get(dedupKey);
    if (lastSeen && timestamp.getTime() - lastSeen < DEDUP_WINDOW_MS) {
        return { valid: false, errors: ['Duplicate event (deduplicated)'] };
    }
    recentEvents.set(dedupKey, timestamp.getTime());

    // Clean up old entries every 1000 events to prevent memory leak
    if (recentEvents.size > 1000) {
        const cutoff = Date.now() - 5000;
        for (const [key, ts] of recentEvents.entries()) {
            if (ts < cutoff) recentEvents.delete(key);
        }
    }

    return {
        valid: true,
        cleaned: {
            user_id: raw.user_id.trim(),
            session_id: raw.session_id.trim(),
            x: Math.round(raw.x),
            y: Math.round(raw.y),
            timestamp,
            page: raw.page || '/',
        },
    };
}

module.exports = { cleanEvent };
